src/main/java/org/opentripplanner/middleware/controllers/api/ApiUserController.java

```diff
@@ -1,28 +1,170 @@
+        //  the parameter and 'withPathParam' is not optional (even with 'withRequired(false))'. For the time being
+        //  appending '?usagePlanId=<id>' works.
+        ApiEndpoint modifiedEndpoint = baseEndpoint
+            // Reveal the actual API key value for the key ID
+            .get(path(ROOT_ROUTE + API_KEY_PATH + API_KEY_ID_PARAM)
+                    .withDescription("Gets the API key value for a given api key ID")
+                    .withPathParam().withName("apiKeyId").withDescription("The api key ID of the API key.").and()
+                    .withResponseType(ApiKey.class),
+                this::getApiKey, JsonUtils::toJson
+            )
+            // Create API key
+            .post(path(API_KEY_PATH)
+                    .withDescription("Creates API key for ApiUser (with optional usagePlanId)")
+                    .withResponseType(persistence.clazz),
+                this::createApiKeyForApiUser, JsonUtils::toJson
+            )
+            // Delete API key
+            .delete(path(API_KEY_PATH + API_KEY_ID_PARAM)
+                    .withDescription("Deletes API key for ApiUser")
+                    .withPathParam().withName("apiKeyId").withDescription("The ID of the API key.").and()
+                    .withResponseType(persistence.clazz),
+                this::deleteApiKeyForApiUser, JsonUtils::toJson
+            )
+
+            // Options response for CORS for the api key path
+            .options(path(API_KEY_PATH), (req, res) -> "");
+
+        // Add the regular CRUD methods after defining the /apikey route.
+        super.buildEndpoint(modifiedEndpoint);
+    }
+
+    /**
+     * HTTP endpoint that reveals the actual API Key value for a given apiKeyId.
+     */
+    private ApiKey getApiKey(Request req, Response res) {
+        Auth0UserProfile requestingUser = Auth0Connection.getUserFromRequest(req);
+        String apiKeyId = HttpUtils.getRequiredParamFromRequest(req, "apiKeyId");
+
+        // User must be admin or have the key in order to view key details.
+        if (isUserAdmin(requestingUser) || userHasKey(requestingUser.apiUser, apiKeyId)) {
+            return new ApiKey(ApiGatewayUtils.getApiKey(apiKeyId));
+        }
+        logMessageAndHalt(req, HttpStatus.FORBIDDEN_403, "User not permitted to view API key.");
+        return null;
+    }
+
+    /**
+     * Shorthand method to determine if an API user exists and has an API key.
+     */
+    private boolean userHasKey(ApiUser user, String apiKeyId) {
+        return user != null &&
+            user.apiKeys
+                .stream()
+                .filter(ApiKey -> apiKeyId.equals(ApiKey.id))
+                .findAny()
+                .orElse(null) != null;
+    }
+
+    /**
+     * Create a new API key and assign it to the provided usage plan. If no usage plan is provided use the default
+     * usage plan instead.
+     */
+    private ApiUser createApiKeyForApiUser(Request req, Response res) {
```

Comment:
I am not really in favor of returning the entire user as a result of this operation. Could that call just return the created key?

Reply:
I'm defering this one to @landonreed

Reply:
It's simpler in terms of API interaction to just return the API user. I think we should still return the entire API user.